### PR TITLE
feat: add SASL  and TLS config for Kafka client

### LINKED ISSUE #4534: Allow users to configure the SASL and TLS for the Kafka client
### What problem does the new feature solve?

This feature will secure Kafka connections by enabling encrypted communication and ensuring that only authorized users can connect.

### What does the feature do?

Allow users to configure the SASL/PLAIN, SASL/SCRAM-SHA-256 SASL/SCRAM-SHA-512 and TLS settings for the Kafka client.

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
close #4534
## What's changed and what's your intention?

Allow users to configure the SASL/PLAIN, SASL/SCRAM-SHA-256, SASL/SCRAM-SHA-512 and TLS for the Kafka client

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

### CODE REVIEW

File: src/common/wal/Cargo.toml [hunk 1]
```diff
@@ -18,6 +18,8 @@ common-telemetry.workspace = true
 futures-util.workspace = true
 humantime-serde.workspace = true
 rskafka.workspace = true
+rustls = { version = "0.23", default-features = false, features = ["ring", "logging", "std", "tls12"] }
+rustls-pemfile = "2.1"
```
Comment (WenyXu) [inline src/common/wal/Cargo.toml:22]: Should we update the these deps in the `servers` crate? @sunng87

Comment (sunng87) [reply src/common/wal/Cargo.toml:22]: I remember it was tonic blocking us from the upgrade. We need to update tonic to 0.12.

Comment (sunng87) [reply src/common/wal/Cargo.toml:22]: I think we can upgrade rustls of servers in another PR. It requires upgrade of proto module as well and could take some time.

**Comment by WenyXu** [comment]
SCRAM-SHA-512 requires https://github.com/dequbed/rsasl/pull/48, https://github.com/influxdata/rskafka/pull/247

File: src/common/wal/src/config/kafka/datanode.rs [hunk 2]
```diff
@@ -39,6 +45,105 @@ pub struct DatanodeKafkaConfig {
...
+impl KafkaClientTls {
+    /// Builds the [`ClientConfig`].
+    pub fn to_tsl_config(&self) -> Result<Arc<ClientConfig>> {
+        let builder = ClientConfig::builder();
+        let mut roots = RootCertStore::empty();
+
+        let root_cert_bytes =
+            std::fs::read(&self.server_ca_cert_path).context(error::ReadFileSnafu {
+                path: &self.server_ca_cert_path,
+            })?;
```
Comment (v0y4g3r) [inline src/common/wal/src/config/kafka/datanode.rs:112]: Avoid synchronous IO in async context. Looks like `KafkaClientTls::to_tsl_config` is invoked from `ClientManager::try_new`.

**Review by zyy17 — COMMENTED** [review]
Most of changes look good to me. Should we need some unit tests to make the auth work ok?

File: src/common/wal/src/config/kafka/datanode.rs [hunk 3]
```diff
@@ -39,6 +45,105 @@ pub struct DatanodeKafkaConfig {
...
+        let config = if let (Some(cert_path), Some(key_path)) =
+            (&self.client_cert_path, &self.client_key_path)
+        {
+            let cert_bytes =
+                std::fs::read(cert_path).context(error::ReadFileSnafu { path: cert_path })?;
+            let client_certs = rustls_pemfile::certs(&mut Cursor::new(cert_bytes))
+                .collect::<std::result::Result<Vec<_>, _>>()
+                .context(error::ReadCertsSnafu { path: cert_path })?;
+
+            let key_bytes = std::fs::read(key_path).unwrap();
```
Comment (v0y4g3r) [inline src/common/wal/src/config/kafka/datanode.rs:133]: Looks like this unwrap is not safe here.

Comment (WenyXu) [reply src/common/wal/src/config/kafka/datanode.rs:133]: Good point

**Comment by WenyXu** [comment]
> Should we need some unit tests to make the auth work ok?

Added in [735e6b1](https://github.com/GreptimeTeam/greptimedb/pull/4536/commits/735e6b13114cb76b2b1f8726b7639f2282eabf91)

**Review by zyy17 — APPROVED** [review]

File: src/common/wal/src/error.rs [hunk 4]
```diff
@@ -24,10 +24,74 @@ pub enum Error {
...
+
+    #[snafu(display("Failed to set client auth cert"))]
+    SetClientAuthCert {
+        #[snafu(source)]
+        error: rustls::Error,
+        #[snafu(implicit)]
+        location: Location,
+    },
+
+    #[snafu(display("Failed to ca certs from system"))]
```
Comment (v0y4g3r) [inline src/common/wal/src/error.rs:88]:
```suggestion
    #[snafu(display("Failed to load ca certs from system"))]
```

**Review by v0y4g3r — APPROVED** [review]
Mostly LGTM

**Pull request merged**